src/main/software/amazon/event/ruler/ComparableNumber.java

```diff
@@ -3,104 +3,122 @@
 
-    public static final BigDecimal TEN_E_SIX = new BigDecimal("1E6"); // to remove decimals
-    public static final long HALF_TRILLION_TEN_E_SIX = new BigDecimal(ComparableNumber.HALF_TRILLION).multiply(TEN_E_SIX).longValueExact();
-
-    private ComparableNumber() {
-    }
+    private ComparableNumber() {}
 
     /**
-     * Generates a hexadecimal string representation of a given decimal string value,
-     * with a maximum precision of 6 decimal places and a range between -5,000,000,000
-     * and 5,000,000,000 (inclusive).
+     * Generates a comparable number string from a given string representation
+     * using numbits representation.
      *
-     * @param str the decimal string value to be converted
-     * @return the hexadecimal string representation of the input value
-     * @throws IllegalArgumentException if the input value has more than 6 decimal places
-     *                                  or is outside the allowed range
+     * @param str the string representation of the number
+     * @return the comparable number string
+     * @throws NumberFormatException if the input isn't a number
+     * @throws IllegalArgumentException if the input isn't a number we can compare
      */
     static String generate(final String str) {
-        final BigDecimal number = JavaBigDecimalParser.parseBigDecimal(str).stripTrailingZeros();
-        if (number.scale() > MAX_DECIMAL_PRECISON) {
-            throw new IllegalArgumentException("Only values upto 6 decimals are supported");
-        }
-
-        final long shiftedBySixDecimals = number.multiply(TEN_E_SIX).longValueExact();
+        final BigDecimal bigDecimal = JavaBigDecimalParser.parseBigDecimal(str);
+        final double doubleValue = bigDecimal.doubleValue();
 
-        // faster than doing bigDecimal comparisons
-        if (shiftedBySixDecimals < -HALF_TRILLION_TEN_E_SIX || shiftedBySixDecimals > HALF_TRILLION_TEN_E_SIX) {
-            throw new IllegalArgumentException("Value must be between " + -ComparableNumber.HALF_TRILLION +
-                    " and " + ComparableNumber.HALF_TRILLION + ", inclusive");
+        // make sure we have the comparable numbers and haven't eaten up decimals values
+        if(Double.isNaN(doubleValue) || Double.isInfinite(doubleValue) ||
+                BigDecimal.valueOf(doubleValue).compareTo(bigDecimal) != 0) {
+            throw new IllegalArgumentException("Cannot compare number : " + str);
         }
+        final long bits = Double.doubleToRawLongBits(doubleValue);
 
-        return longToBase64Bytes(shiftedBySixDecimals + HALF_TRILLION_TEN_E_SIX);
+        // if high bit is 0, we want to xor with sign bit 1 << 63, else negate (xor with ^0). Meaning,
+        // bits >= 0, mask = 1000000000000000000000000000000000000000000000000000000000000000
+        // bits < 0,  mask = 1111111111111111111111111111111111111111111111111111111111111111
+        final  long mask = ((bits >>> 63) * 0xFFFFFFFFFFFFFFFFL) | (1L  << 63);
```

Comment:
Raph Levien proposed an even nicer way on Mastodon (not yet in Quamina afaik).
`(bits >> 63) | (1L << 63)` is sufficient. For negative numbers, the sign bit will be set. `>>` is sign extending and will fill all the left bits with the sign bit. No need for another constant.

Reply:
sweet. Thanks. If you have the link to the tweet (toot?) from Raph, I'd love to link it for attribution.

Reply:
https://mastodon.online/@raph/113071041069390831